### PR TITLE
Port to Qt 5.12

### PR DESCRIPTION
**Pull request by dobey**
This is a big change to build uitk against newer version of Qt. As in the
past, some new tests now result in crashes or odd failures due to changes
in Qt, so have had to be skipped, though others have been fixed and the
toolkit behaves fine at runtime in testing.

### CODE REVIEW

**Review by mariogrip — APPROVED** [review]
LGTM :+1: 

Also maybe we should make a short docs page about api changes in this? Its minimal but there is  a api change.

File: tests/unit/visual/tst_tabs_with_repeater.11.qml [hunk 1]
```diff
@@ -184,10 +185,12 @@ MainView {
         }
 
         function test_twoRepeaters() {
+            skip("FIXME: Fails on Qt 5.12");
             var tabsModel = twoRepeaters.__model;
             var secondRepeaterModel = secondRepeater.model;
 
             compare(tabsModel.count, firstRepeater.count + secondRepeater.count, "Incorrect number of tabs in TabBar");
+            wait(250);
```
Comment (UniversalSuperBox) [inline tests/unit/visual/tst_tabs_with_repeater.11.qml]: Adding waits with skips seems like not what we want

File: tests/unit/visual/tst_tabs_with_repeater.11.qml [hunk 2]
```diff
@@ -133,7 +134,7 @@ MainView {
             inputModel.move(3, 0, 1);
             inputModel.move(1, 3, 1);
             // wait few miliseconds
-            wait(50);
+            wait(250);
```
Comment (UniversalSuperBox) [inline tests/unit/visual/tst_tabs_with_repeater.11.qml]: Another wait with skip

File: tests/unit/bottomedge/tst_bottomedge.cpp [hunk 3]
```diff
@@ -855,15 +855,14 @@ private Q_SLOTS:
     {
         QFETCH(QString, document);
 
-        QScopedPointer<BottomEdgeTestCase> test(new BottomEdgeTestCase(document));
+        auto test = new BottomEdgeTestCase(document);
```
Comment (UniversalSuperBox) [inline tests/unit/bottomedge/tst_bottomedge.cpp]: Has Qt's memory handling somehow improved to make the ScopedPointer no longer necessary, or is there another reason to use auto instead?

File: src/imports/Components/1.3/Captions.qml [hunk 4]
```diff
@@ -124,22 +124,16 @@ ColumnLayout {
     }
 
     Label {
+        Layout.fillWidth: true
```
Comment (UniversalSuperBox) [inline src/imports/Components/1.3/Captions.qml:127]: Is this an API break?

File: src/imports/Components/1.2/Captions.qml [hunk 5]
```diff
@@ -124,22 +124,16 @@ ColumnLayout {
     }
 
     Label {
+        Layout.fillWidth: true
```
Comment (UniversalSuperBox) [inline src/imports/Components/1.2/Captions.qml:127]: Same as 1.3 Caption, is this an API break?

File: debian/rules [hunk 6]
```diff
@@ -7,18 +7,18 @@ export CFLAGS := $(shell dpkg-buildflags --get CFLAGS) $(shell dpkg-buildflags -
...
 
 ifneq ($(DEB_BUILD_GNU_TYPE),$(DEB_HOST_GNU_TYPE))
 	# Cross compiling
 	CONFIGURE_OPTS += "CONFIG+=no_docs" "CONFIG+=cross_compile"
 endif
 
 # Skip tests on the archs they are known to be flaky  with current configuration
-testskip_architectures := armhf arm64 powerpc ppc64el s390x
-testskip_series := 
+testskip_architectures := armhf powerpc ppc64el s390x
```
Comment (UniversalSuperBox) [inline debian/rules:21]: Tests still fail a lot on armhf then?

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
Just a few things to say

File: tests/unit/mousefilters/tst_mousefilterstest.cpp [hunk 7]
```diff
@@ -851,7 +851,7 @@ private Q_SLOTS:
         QCOMPARE(pressed.count(), 1);
         QCOMPARE(released.count(), 1);
         QCOMPARE(clicked.count(), 1);
-        QCOMPARE(positionChanged.count(), 30);
+        QCOMPARE(positionChanged.count(), 31);
```
Comment (peat-psuwit) [inline tests/unit/mousefilters/tst_mousefilterstest.cpp:854]: Do you know why this is needed?

File: components.api [hunk 8]
```diff
@@ -289,7 +289,7 @@ Ubuntu.Components.Popups.ComposerSheet 1.3: SheetBase
 Ubuntu.Layouts.ConditionalLayout 1.0 0.1 ULConditionalLayout: QtObject
     default property Component layout
     property string name
-    property QQmlBinding when
+    property bool when
```
Comment (peat-psuwit) [inline components.api:292]: Although it probably doesn't affect QML users, why is this changed?

File: debian/rules [hunk 9]
```diff
@@ -28,16 +28,13 @@ override_dh_auto_configure:
...
-	echo $(DEB_HOST_SERIES)
-	echo $(testskip_series)
-ifneq ($(DEB_HOST_SERIES),$(findstring $(DEB_HOST_SERIES), $(testskip_series)))
-ifneq ($(DEB_HOST_ARCH),$(findstring $(DEB_HOST_ARCH), $(testskip_architectures)))
-ifeq (,$(findstring nocheck,$(DEB_BUILD_OPTIONS)))
 	mkdir -p debian/tmp/home/run
 	chmod +x tests/xvfb.sh
-	tests/xvfb.sh dh_auto_test || exit $?
-endif
+ifeq (,$(filter nocheck,$(DEB_BUILD_OPTIONS)))
```
Comment (peat-psuwit) [inline debian/rules]: Nitpick, but I guess we can move the preparation under `nocheck` check.

```Makefile
ifeq (,$(filter nocheck,$(DEB_BUILD_OPTIONS)))
	mkdir -p debian/tmp/home/run
	chmod +x tests/xvfb.sh
ifneq (,$(filter $(DEB_HOST_ARCH),$(testskip_architectures)))
<....>
```

File: src/imports/Components/1.2/Tabs.qml [hunk 10]
```diff
@@ -414,7 +414,7 @@ PageTreeNode {
 
               Inspired from http://perfectionkills.com/instanceof-considered-harmful-or-how-to-write-a-robust-isarray/
               */
-            if (Object.prototype.toString.call(repeater.model) !== "[object Object]") {
+            if (!repeater.model || repeater.model.toString() !== "[object Object]") {
```
Comment (peat-psuwit) [inline src/imports/Components/1.2/Tabs.qml]: Is this change due to the previous `Object` prototype breakage? If so I would suggest that this (and all instances in other files) be reverted/dropped for more predictable behavior as this code relies on `toString()` being from `Object`.

For example, a code in another file checks for `model.toString() === "[object Array]"`, however (from Chromium 87):
image

File: tests/unit/visual/tst_listitem.12.qml [hunk 11]
```diff
@@ -649,7 +649,7 @@ Item {
             return [
                 {tag: "No actions", item: highlightTest, x: centerOf(highlightTest).x, y: centerOf(highlightTest).y, pressed: false},
                 {tag: "Leading/trailing actions", item: testItem, x: centerOf(testItem).x, y: centerOf(testItem).y, pressed: true},
-                {tag: "Active component content", item: controlItem, x: units.gu(1), y: units.gu(1), pressed: true},
+                {tag: "Active component content", item: controlItem, x: units.gu(1), y: units.gu(1), pressed: false},
```
Comment (peat-psuwit) [inline tests/unit/visual/tst_listitem.12.qml:652]: Are you sure this is correct?

File: tests/unit/visual/tst_picker.11.qml [hunk 12]
```diff
@@ -288,10 +289,10 @@ Item {
             waitPickerScrolling();
 
             compare(linearList.count, 14, "bad removal from linearList")
-            compare(linearDynPicker.selectedIndex, 4, "bad index of linearList");
+            compare(linearDynPicker.selectedIndex, 5, "bad index of linearList");
```
Comment (peat-psuwit) [inline tests/unit/visual/tst_picker.11.qml]: Why this is changed.

File: tests/unit/visual/tst_tabs.13.qml [hunk 13]
```diff
@@ -116,7 +116,7 @@ MainView {
     Component {
         id: pageComponentNoFlick
         Page {
-            title: "Loaded page without flickable"
+            header: PageHeader { title: "Loaded page without flickable" }
```
Comment (peat-psuwit) [inline tests/unit/visual/tst_tabs.13.qml:119]: Is the old code supposed to work?

**Review by peat-psuwit — CHANGES_REQUESTED** [review]
I would prefer to see multiple commits of manageable size. Otherwise, it's a bit hard to review. Preferably each commit should answer its own questions.

Also, changes to the test case should come with more explanation. I didn't comment on all of them, but there're several unexplained test case changes other than simply skipping/unskipping or verify() - compare() fixes.

File: src/imports/Components/1.2/Captions.qml (hunk 5 above)
Comment (dobey) [reply src/imports/Components/1.2/Captions.qml:127]: No.

File: src/imports/Components/1.3/Captions.qml (hunk 4 above)
Comment (dobey) [reply src/imports/Components/1.3/Captions.qml:127]: No. It's using the correct property as anchors cannot be used on elements inside a `*Layout` component.

File: debian/rules (hunk 6 above)
Comment (dobey) [reply debian/rules:21]: Unfortunately, yes. It seems even though things are much less flaky on amd64 and arm64, they are still pretty flaky on armhf.

File: tests/unit/visual/tst_tabs.13.qml (hunk 13 above)
Comment (dobey) [reply tests/unit/visual/tst_tabs.13.qml:119]: I don't know. It results in several deprecation warnings, as well as resulting in the height being different causing a test failure, without the change.

File: tests/unit/visual/tst_listitem.12.qml (hunk 11 above)
Comment (dobey) [reply tests/unit/visual/tst_listitem.12.qml:652]: Not totally, but the test doesn't quit make sense to me logically either. However, the data item immediately following this one is using the same `item` and also has `pressed` as `false` which suggests that this change is correct.

File: tests/unit/mousefilters/tst_mousefilterstest.cpp (hunk 7 above)
Comment (dobey) [reply tests/unit/mousefilters/tst_mousefilterstest.cpp:854]: Not exactly, but I presume some change in Qt results in the `mousePress()` call before the for loop to move the mouse 30 times, also results in a `positionChanged` event, given the moues area is clear at that point.

File: components.api (hunk 8 above)
Comment (dobey) [reply components.api:292]: `QQmlBinding` was coming from private API. Something had changed in Qt such that the code for the layouts plugin would crash when calling `evaluate()->toBool()` on the binding. Since it expects a bool there anyway, and Qt already internally does appropriate binding, it was better to get rid of the dependency on the private API and just use a bool instead.

File: tests/unit/visual/tst_tabs_with_repeater.11.qml (hunk 1 above)
Comment (dobey) [reply tests/unit/visual/tst_tabs_with_repeater.11.qml]: Fixed.

File: tests/unit/visual/tst_tabs_with_repeater.11.qml (hunk 2 above)
Comment (dobey) [reply tests/unit/visual/tst_tabs_with_repeater.11.qml]: Also fixed.

File: src/imports/Components/1.2/Tabs.qml (hunk 10 above)
Comment (dobey) [reply src/imports/Components/1.2/Tabs.qml]: Reverted this and others. Thanks.

File: tests/unit/visual/tst_picker.11.qml (hunk 12 above)
Comment (dobey) [reply tests/unit/visual/tst_picker.11.qml]: I guess the `Object` issues affect this. Reverted now.

File: debian/rules (hunk 9 above)
Comment (dobey) [reply debian/rules]: Done.

File: tests/unit/bottomedge/tst_bottomedge.cpp (hunk 3 above)
Comment (dobey) [reply tests/unit/bottomedge/tst_bottomedge.cpp]: I'm not sure why the ScopedPointer is used in these tests exactly, as I don't think it should be necessary. These changes were made due to the test crashing. However, reverting this seems to work again (I am not sure why). So assuming the new build succeeds in CI, I think we can consider this resolved too.

File: components.api (hunk 8 above)
Comment (peat-psuwit) [reply components.api:292]: Ok, so the old code was crashing.

File: src/imports/Components/Pickers/1.2/Picker.qml [hunk 14]
```diff
@@ -221,7 +221,7 @@ StyledItem {
             target: loader.item
             property: "currentIndex"
             value: picker.selectedIndex
-            when: loader.completed && (picker.selectedIndex > 0)
+            when: picker.selectedIndex > 0
```
Comment (peat-psuwit) [inline src/imports/Components/Pickers/1.2/Picker.qml:224]: Why don't we need to check for `loader.completed`?

File: tests/unit/visual/tst_combobutton.11.qml [hunk 15]
```diff
@@ -178,9 +178,7 @@ Item {
         }
 
         function test_flickRectCombo() {
-            // skip the test temporarily due to flakyness on powerpc target
             // https://bugs.launchpad.net/ubuntu-ui-toolkit/+bug/1315244
-            skip("The test is flaky on powerpc target");
```
Comment (peat-psuwit) [inline tests/unit/visual/tst_combobutton.11.qml]: Maybe you want to also remove the link to the bug too? Also, maybe you want to do the architecture-specific skip.

Comment (dobey) [reply tests/unit/visual/tst_combobutton.11.qml]: Yeah, link removed. Don't think we need to do a skip for powerpc here, since we don't build on it anyway. So let's leave that for when we get the rename to lomiri done and we resolve any differences between here and `lomiri-ui-toolkit` repo on gitalb, for when debian needs ppc builds.

File: src/imports/Components/Pickers/1.2/Picker.qml (hunk 14 above)
Comment (dobey) [reply src/imports/Components/Pickers/1.2/Picker.qml:224]: It creates a binding loop as the `Binding` component here is a child of the `Loader` and it depends on `loader.item` which is undefined until the loader has completed loading it, IIRC.

**Review by peat-psuwit — APPROVED** [review]
As
- all remaining questions are test-related (plus a minor code one, but I'll give a pass on that)
- this PR is the last blocking rootfs generation
- this has been tested as a part of the bigger Qt 5.12 migration, and @dobey confirms that it works at runtime

I'll approve the PR (after this last commit is built successfully) to unblock the rootfs generation. But please leave the questions unresolved so that we can re-visit them later more easily.

**Comment by dobey** [comment]
> * this PR is the last blocking rootfs generation

Not exactly. There are plenty more PRs to come, but they will most all depend on this being merged first, as things have to build against the newly built versions of things for correct ABI linkage. With this unblocked, I can get the remaining PRs opened, as most of them are small and should go more smoothly (unity8 one is I think pretty large again though).

**Comment by peat-psuwit** [comment]
s/rootfs generation/Qt 5.12 transition/ :D

**Comment by dobey** [comment]
> Also maybe we should make a short docs page about api changes in this? Its minimal but there is a api change.

There is no API change.

**Review by UniversalSuperBox — APPROVED** [review]

File: src/imports/Components/Pickers/1.2/Picker.qml (hunk 14 above)
Comment (peat-psuwit) [reply src/imports/Components/Pickers/1.2/Picker.qml:224]: Hmm, I'm not sure why it caused a binding loop. `Binding` should be able to handle `undefined` `target` just fine. Apparently `completed` is a bit more than the `Loader` having loaded the component.

Comment (dobey) [reply src/imports/Components/Pickers/1.2/Picker.qml:224]: With `loader.completed` there, it results in the following warning:
```
file:///build/ubuntu-ui-toolkit-1.3.2192~0ubports2/qml/Ubuntu/Components/Pickers/1.2/Picker.qml:220:9: QML Binding: Binding loop detected for property "when"
```
I'm pretty sure this is because the `Binding` is a direct child of `loader` here.
